File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {

        unordered_map<char, int> counter;
        int l = 0, ans = 0;
        int n = s.size();
        for (int r = 0; r < n; ++r) {
            counter[s[r]]++;
            if (counter.size() < 3) continue;
            while (counter[s[l]] > 1) {
                counter[s[l]]--;
                l++;
            }
            ans += l + 1;
        }



        return ans;
        
    }
};

File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        
        n = len(s)
        ans = 0
        l = 0
        counter = collections.defaultdict(int)
        for r in range(n):
            counter[s[r]] += 1
            if len(counter) < 3:
                continue
            while counter[s[l]] > 1:
                counter[s[l]] -= 1
                l += 1
            ans += l + 1

        return ans
